cluster-operator/src/main/java/io/strimzi/operator/cluster/Main.java

```diff
@@ -77,39 +88,66 @@ public static void main(String[] args) {
         // Verticle.stop() methods are not executed if you don't call Vertx.close()
         // Vertx registers a shutdown hook for that, but only if you use its Launcher as main class
         Runtime.getRuntime().addShutdownHook(new Thread(new ShutdownHook(vertx)));
-        
+
+        // Setup Micrometer Metrics provider
+        MetricsProvider metricsProvider = new MicrometerMetricsProvider();
+
         KubernetesClient client = new KubernetesClientBuilder().build();
 
-        maybeCreateClusterRoles(vertx, config, client).onComplete(crs -> {
-            if (crs.succeeded())    {
-                PlatformFeaturesAvailability.create(vertx, client).onComplete(pfa -> {
-                    if (pfa.succeeded()) {
-                        LOGGER.info("Environment facts gathered: {}", pfa.result());
-
-                        run(vertx, client, pfa.result(), config).onComplete(ar -> {
-                            if (ar.failed()) {
-                                LOGGER.error("Unable to start operator for 1 or more namespace", ar.cause());
-                                System.exit(1);
-                            }
-                        });
-                    } else {
-                        LOGGER.error("Failed to gather environment facts", pfa.cause());
+        maybeCreateClusterRoles(vertx, config, client)
+                .compose(i -> startHealthServer(vertx, metricsProvider))
+                .compose(i -> createPlatformFeaturesAvailability(vertx, client))
+                .compose(pfa -> deployClusterOperatorVerticles(vertx, client, metricsProvider, pfa, config))
```

Comment:
I was wondering if the health server should be the last one to be started when the cluster operator verticle is up ad running so that the K8S platform gets the readiness of the operator when it's really started.

Reply:
The webserver just returns HTTP OK all the time. I does not say anything about the ClusterOperator instance(s). But having it before the `deployClusterOperatorVerticles` will be important while implementing a leadership eelction into the operator (which is actually how I found this bug), where the `ClusterOperator` start will be possibly delayed for when the operator instance becomes a leader.

Reply:
> I does not say anything about the ClusterOperator instance(s)

Not for this PR but maybe this is a problem, or? Should we have a readiness returning OK only when the CO is really working and "watching" for Kafka related custom resources?

> But having it before the deployClusterOperatorVerticles will be important while implementing a leadership eelction

Can you elaborate a little bit why the HTTP server is useful to run earlier when leadership election will be in place?

Reply:
> Can you elaborate a little bit why the HTTP server is useful to run earlier when leadership election will be in place?

When you have a leadership election in place and run multiple instances of the operator, only one of them will be a leader and will _operate_. The others need to sit and wait if they ever become leaders. And during this time, the health checks need to pass as well. Otherwise Kubernetes will see these processes as failing and will crash loop them. That beats the purpose of the leader election for example as discussed in #7158.

> Not for this PR but maybe this is a problem, or? Should we have a readiness returning OK only when the CO is really working and "watching" for Kafka related custom resources?

Well, we do not do that today and I don't think anyone has any easy solution for how this would work (I.e. how to check that the verticles are really working). So I would say we should get there when we actually have something for that.

Reply:
>  And during this time, the health checks need to pass as well. Otherwise Kubernetes will see these processes as failing and will crash loop them.

Ok it makes more sense but does it make sense to have the `liveness` passing but not the `readiness` until the leader is actually elected? Just for reference in the future even referring to your second comment.

Reply:
I would say that it is ready in both cases. Just when you are not a leader, readiness does not mean you are operating resources. From my experience, having something not ready for a long time triggers all kind of monitoring tools etc.. So I would avoid it.

Reply:
I think that "readiness" concept for an operator is kind of ... not well defined.
For a "normal" application exposed via HTTP for example, the readiness is important for avoiding that request are coming through the corresponding K8S `Service` to you: AFAIK, only when your probe reports that you are "ready", the request is forwarded through the `Service` to your application.
For an operator, "readiness" sounds to me that the operator is "watching" custom resources or able to do a periodic reconcile, so it's ready to handle a custom resource creation/update/deletion.

Reply:
From my point of view -> readiness should mean whether the operator is ready. When it is ready as a _warm stand-by_  and waiting to become a leader, it is from my point of view ready as well even when it does not watch resources. That is at the end the point of the leadership election to have something what is _ready_ to take over immediately when the other instance of the operator fails. I'm not sure the HTTP example is easily applicable here because it does not describe the same concept of _warm stand-by_.

I do not think the liveness versus readiness concept is used in general to indicate this - for example for databases which often use the active / passive replica concepts. I haven't seen anything what would be not-ready intentionally for weeks.